webrender/src/clip_stack.rs

```diff
@@ -0,0 +1,178 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashMap;
+use euclid::{Rect, Matrix4D};
+use gpu_store::{GpuStore, GpuStoreAddress};
+use internal_types::DeviceRect;
+use prim_store::{ClipData, GpuBlock32, PrimitiveClipSource, PrimitiveStore};
+use prim_store::{CLIP_DATA_GPU_SIZE, MASK_DATA_GPU_SIZE};
+use tiling::StackingContextIndex;
+use util::TransformedRect;
+use webrender_traits::{AuxiliaryLists, ImageMask};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub struct ClipAddressRange {
+    pub start: GpuStoreAddress, // start GPU address
+    pub count: u32, // number of items, not bytes
+}
+
+type ImageMaskIndex = u16;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+pub struct MaskCacheKey {
+    pub layer_id: StackingContextIndex,
+    pub clip_range: ClipAddressRange,
+    pub image: Option<GpuStoreAddress>,
+}
+
+impl MaskCacheKey {
+    pub fn empty(layer_id: StackingContextIndex) -> MaskCacheKey {
+        MaskCacheKey {
+            layer_id: layer_id,
+            clip_range: ClipAddressRange {
+                start: GpuStoreAddress(0),
+                count: 0,
+            },
+            image: None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct MaskCacheInfo {
+    pub key: MaskCacheKey,
+    // this is needed to update the ImageMaskData after the
+    // ResourceCache allocates/load the actual data
+    // will be simplified after the TextureCache upgrade
+    pub image: Option<ImageMask>,
+    pub device_rect: DeviceRect,
```

Comment:
One problem that I see with this change is the computation of `device_rect`. Since layers are supposed to move, this rectangle needs to be re-computed, but there is not tracking of the layers at this point. Is there an example where the layers state is being cached and updated upon any movement? Alternatively, I can re-compute the rectangle every frame.